src/wasm-type.h

```diff
@@ -97,15 +97,19 @@ class Type {
   // │ f64         ║ x │   │ x │ x │   F   │ │  V_ector
   // │ v128        ║ x │   │ x │ x │     V │ ┘
   // ├─────────────╫───┼───┼───┼───┤───────┤
-  // │ funcref     ║ x │   │ x │ x │ f     │ ┐ Ref
-  // │ externref   ║ x │   │ x │ x │       │ │  f_unc
-  // │ nullref     ║ x │   │ x │ x │       │ │
-  // │ exnref      ║ x │   │ x │ x │       │ │
+  // │ funcref     ║ x │   │ x │ x │ f  n  │ ┐ Ref
+  // │ externref   ║ x │   │ x │ x │ f? n  │ │  f_unc, n_ullable
+  // │ anyref      ║ x │   │ x │ x │ f? n  │ │ ┐
+  // │ eqref       ║ x │   │ x │ x │    n  │ │ │ TODO (GC)
+  // │ i31ref      ║ x │   │ x │ x │       │ │ ┘
```

Comment:
Seems there is a nullability exception for `i31ref`, but I might be mistaken?

Reply:
Where do you see this?

Reply:
[Here](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#reference-types) everything is a `(ref null X)` except `i31ref`.

Reply:
Hmm, that definitely says that `i31ref` is an abbreviation for a non-nullable reference to i31 (i.e. `(ref i31)`), but it looks like you could also construct a `(ref null i31)` (which would be encoded as `-0x14 -0x17`).

Reply:
Hmm, so in this case it would be correct that `i31ref` is non-nullable, but we are missing a way to express `(ref null i31)`, perhaps as another compound kind of `Rtt` with just a `bool` in the union. Very strange :)

Reply:
Or, well, since that's a type, not even in `Rtt`, but as a special kind of `TypeInfo`. Even stranger.

Reply:
Thinking about these abbreviations more, I realized that when we canonicalize types we will need to be able to recognize constructed types like `(ref null any)` and canonicalize them to their abbreviations like `anyref`. Arguably we should not even use the abbreviated types except in parsing, but that seems like it would be a riskier change to make since it would be a departure from the current specification.

Reply:
Yeah `i31` is basically the same as `(array ...)` or `(struct ...)` as far as the Type grammar is concerned (i.e. it is a HeapType), so it would make sense to add a `TypeInfo` variant for it.

Reply:
I was assuming that we can just use the short forms of the types and translate to them during parsing, since these are a little easier to look at. But as you say, the other way around makes a lot more sense if the mapping has corner cases. If so,  it might be good to introduce a `HeapType` as mentioned in https://github.com/WebAssembly/binaryen/pull/3076#pullrequestreview-473927409, but not just as an `enum` but as a union that can then be used by both `TypeInfo` and `Rtt`m while also removing `anyref`, `eqref` and `i31ref` from basic types since these are really just aliases. Hmm...

Reply:
I would be fine with that if that's what you'd like to do 👍

Reply:
Alright, refactored to use a common `HeapType`, which replaces the `union` in `TypeInfo` to describe any compound ref type, and can be easily used by `Rtt` as well. Turned out that this even simplifies quite a few other parts, like other constructors taking a `HeapType` already can now automatically deal with anything used to construct as a `HeapType` (i.e. HeapType::Kind, Signature, Struct and Array), the distinction into `tuple`, `ref` and `rtt` is nice to represent `nullable` only applicable to `Ref`s, and we can even canonicalize basic heap types to just `anyref` etc. but putting these into `indices` (I think, hope this works).